ChildProject/pipelines/metrics.py

```diff
@@ -80,46 +120,266 @@ def __init__(
+        
+        return row[1]
+                
+    
     def extract(self):
-        pass
+        """from the initiated self.metrics, compute each row metrics (handles threading)
+        Once the Metrics class is initialized, call this function to extract the metrics and populate self.metrics
+        
+        :return: DataFrame of computed metrics
+        :rtype: pandas.DataFrame
+        """
+        if self.threads == 1:
+            self.metrics = pd.DataFrame(
+                [self._process_unit(row) for row in self.metrics.iterrows()]
+            )
+        else:
+            with mp.Pool(
+                processes=self.threads if self.threads >= 1 else mp.cpu_count()
+            ) as pool:
+                self.metrics = pd.DataFrame(
+                    pool.map(self._process_unit, self.metrics.iterrows())
+                )
+        if self.period:
+            self.metrics['period_start'] = self.metrics['period_start'].dt.strftime('%H:%M:%S')
+            self.metrics['period_end'] = self.metrics['period_end'].dt.strftime('%H:%M:%S')
+        return self.metrics
 
-    def retrieve_segments(self, sets: List[str], unit: str):
-        annotations = self.am.annotations[self.am.annotations[self.by] == unit]
+    def retrieve_segments(self, sets: List[str], row: str):
+        """from a list of sets and and a row identifying the unit computed, return the relevant annotation segments
+        
+        :param sets: List of annotation sets to keep
+        :type sets: List[str]
+        :param row: Series storing the unit to compute information
+        :type row: pandas.Series
+        :return: relevant annotation DataFrame and index DataFrame
+        :rtype: (pandas.DataFrame , pandas.DataFrame)
+        """
+        annotations = self.am.annotations[self.am.annotations[self.by] == row[self.by]]
         annotations = annotations[annotations["set"].isin(sets)]
-
+        
         if self.from_time and self.to_time:
-            annotations = self.am.get_within_time_range(
-                annotations, self.from_time, self.to_time, errors="coerce"
-            )
+            if self.period:
+                st_hour = row["period_start"]
+                end_hour = row["period_end"]
+                intervals = time_intervals_intersect(TimeInterval(self.from_time,self.to_time),TimeInterval(st_hour,end_hour))
+                matches = pd.concat([self.am.get_within_time_range(annotations,i) for i in intervals],ignore_index =True) if intervals else pd.DataFrame()
+            else:
+                matches = self.am.get_within_time_range(
+                annotations, TimeInterval(self.from_time,self.to_time))
+        elif self.period:
+            st_hour = row["period_start"]
+            end_hour = row["period_end"]
+            matches = self.am.get_within_time_range(
+                    annotations, TimeInterval(st_hour,end_hour))
+        else:
+            matches = annotations
 
         try:
```

Comment:
What are `segments`? Do exceptions here correspond to "no segments for these annotations"?

Reply:
annotations represent a period of the audio file that is annotated (e.g. a row of annotations having set = 'its' , range_onset = '0' , range_offset = '3600,000' is the first hour of the recording annotated in the its set)
Segments are the actual vocalizations, so if you do `get_segments` on the example of annotation above, it will return a DataFrame containing all the vocalizations (segment_onset/offset, speaker_type [etc](https://childproject.readthedocs.io/en/latest/format.html#id10) ) that were detected by lena it the first hour of the audio.
And yes, the exception is for when no vocalizations exist for this anotated period.

Reply:
Got it, thanks!

Could catching all exceptions lead to problems here? I am not assuming this is necessarily a problem but maybe worth a comment on why it is a good idea in this case?

Reply:
I think you are right, it is probably not justified to catch exceptions for that particular case, the function should be able to have an empty dataframe output there.
(e57fdb2bd)